hashes/poseidon/src/lib.rs

```diff
@@ -51,15 +51,31 @@ impl<F: ScalarField, const T: usize, const RATE: usize> PoseidonState<F, T, RATE
         assert!(inputs.len() < T);
         let offset = inputs.len() + 1;
 
-        self.s[0] =
-            gate.sum(ctx, inputs.iter().map(|a| Existing(*a)).chain([Constant(pre_constants[0])]));
-
+        // Explanation of what's going on: before each round of the poseidon permutation,
+        // two things have to be added to the state: inputs (the absorbed elements) and
+        // preconstants. Imagine the state as a list of T elements, the first of which is
+        // the capacity:  |--cap--|--el1--|--el2--|--elR--|
+        // - A preconstant is added to each of all T elements (which is different for each)
+        // - The inputs are added to all elements starting from el1 (so, not to the capacity),
+        //   to as many elements as inputs are available.
+        // - To the first element for which no input is left (if any), an extra 1 is added.
+
+        // ORIGINAL CODE: CAUSED PROBLEMS
+        // self.s[0] =
+        //     gate.sum(ctx, inputs.iter().map(|a| Existing(*a)).chain([Constant(pre_constants[0])]));
+
+        // adding preconstant to the distinguished capacity element (only one)
+        // REPLACED BY THIS
+        self.s[0] = ctx.load_witness(*self.s[0].value() + pre_constants[0]);
```

Comment:
This would load it entirely as witness, without any constraints. 
Instead you should do `self.s[0] = gate.add(ctx, self.s[0], Constant(pre_constants[0]));`

Reply:
Thank you so much! I found the logical flaw but failed to constrain it on the circuit level. I have now pushed a corrected version according to your suggestion.